src/transformers/models/jukebox/modeling_jukebox.py

```diff
@@ -0,0 +1,3133 @@
+            3: lambda d: [1, 4][d % 2],  # Alternate row and last column
+            4: lambda d: [1, 5][d % 2],  # Alternate row and last k columns
+            5: lambda d: [1, 4, 1, 1][d % 4],  # Alternate row, last column, row, row
+            6: lambda d: [1, 2, 3, 6][d % 4],
+            7: lambda d: [*[1, 2, 3] * 5, 6][d % 16],
+            8: lambda d: [1, 2, 3, 1, 2, 3, 1, 2, 3, 6][d % 10],  # Used by separated_enc_dec model with lyrics
+            9: lambda d: [1, 2, 3, 0][d % 4],
+            10: lambda d: [*[1, 2, 3, 1, 2, 3, 1, 2, 3], *[1, 2, 3, 1, 2, 3, 1, 2, 3, 6] * 7][
+                d % 79
+            ],  # Used by large separated_enc_dec model with lyrics
+            11: lambda d: [6, 6, 0][d % 3] if d % 16 == 15 else [1, 2, 3][d % 3],
+            12: lambda d: [7, 7, 0][d % 3]
+            if d % 16 == 15
+            else [1, 2, 3][d % 3],  # Used by single_enc_dec model with lyrics
+        }[attn_order]
+
+        def attn_block(d):
+            return JukeboxBlock(
+                width=width,
+                n_ctx=n_ctx,
+                num_heads=num_heads,
+                attn_dropout=attn_dropout,
+                resid_dropout=resid_dropout,
+                afn=afn,
+                scale=scale,
+                mask=mask,
+                zero_out=zero_out if attn_func(d) != 6 else True,
+                init_scale=init_scale,
+                res_scale=res_scale,
+                m_attn=m_attn,
+                m_mlp=m_mlp,
+                attn_func=attn_func(d),
+                blocks=blocks,
+                spread=spread,
+                encoder_dims=encoder_dims,
+                lyric_enc_len=lyric_enc_len,
+            )
+
+        self._attn_mods = nn.ModuleList()
+        for d in range(n_depth):
+            self._attn_mods.append(attn_block(d))
+
+        self.saved_attn_weights = []
+
+    def set_record_attn(self, record_attn):
+        """
+        Arguments:
+            record_attn (bool or set): Makes forward prop dump self-attention
+                softmaxes to self.saved_attn_weights. Either a set of layer indices indicating which layers to store,
+                or a boolean value indicating whether to dump all.
+        """
+
+        def _should_record_attn(layer_idx):
+            if isinstance(record_attn, bool):
+                return record_attn
+            return layer_idx in record_attn
+
+        for i, layer in enumerate(self._attn_mods):
+            layer.attn.record_attn = _should_record_attn(i)
+        if record_attn:
+            assert self.saved_attn_weights == []
+            for layer in self._attn_mods:
+                assert layer.attn.w is None
+        else:
+            self.saved_attn_weights = []
+            for layer in self._attn_mods:
+                layer.attn.w = None
+
+    def forward(self, hidden_states, lyric_encoder_states=None, sample=False, fp16=False, fp16_out=False):
+        if fp16:
+            hidden_states = hidden_states.half()
```

Comment:
this should be done automatically - don't htink we need to do it ourselves here no?

Reply:
`hidden_states.dtype` should carry the information of what dtype is needed here -> no need to have a `fp16` input IMO